From david-b@pacbell.net Fri Dec  8 12:50:35 2006
From: Wojtek Kaniewski <wojtekka@toxygen.net>
Subject: USB: at91_udc: Cleanup variables after failure in usb_gadget_register_driver()
Date: Friday 08 December 2006 3:23 am
To: linux-usb-devel@lists.sourceforge.net
Cc: andrew@sanpeople.com
Message-Id: <200612080939.57496.david-b@pacbell.net>

This patch zeroes some variables when usb_gadget_register_driver()
fails. gadgetfs does a dummy registration to get the name of the USB
driver and then waits for user-land driver. If someone plugs the cable
in the meantime, bad things happen, because at91_udc has been left in
inconsistent state.

Signed-off-by: Wojtek Kaniewski <wojtekka@toxygen.net>
Acked-by: David Brownell <dbrownell@users.sourceforge.net>
Cc: Andrew Victor <andrew@sanpeople.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/gadget/at91_udc.c |    4 ++++
 1 file changed, 4 insertions(+)

--- gregkh-2.6.orig/drivers/usb/gadget/at91_udc.c
+++ gregkh-2.6/drivers/usb/gadget/at91_udc.c
@@ -1601,6 +1601,10 @@ int usb_gadget_register_driver (struct u
 	if (retval) {
 		DBG("driver->bind() returned %d\n", retval);
 		udc->driver = NULL;
+		udc->gadget.dev.driver = NULL;
+		udc->gadget.dev.driver_data = NULL;
+		udc->enabled = 0;
+		udc->selfpowered = 0;
 		return retval;
 	}
 
